python/ray/experimental/workflow/storage.py

```diff
@@ -0,0 +1,94 @@
+    local/shared filesystems.
+    """
+
+    def __init__(self, workflow_root_dir: Union[pathlib.Path, str]):
+        self._workflow_root_dir = pathlib.Path(workflow_root_dir)
+
+    def _write_atomic(self, writer, mode, obj, path: pathlib.Path, overwrite):
+        # TODO(suquark): race conditions like two processes writing the
+        # same file is still not safe. This may not be an issue, because
+        # in our current implementation, we only need to guarantee the
+        # file is either fully written or not existing.
+        fullpath = self._workflow_root_dir / path
+        if fullpath.exists() and not overwrite:
+            raise FileExistsError(fullpath)
+        tmp_new_filename = fullpath.with_suffix(fullpath.suffix + "." +
+                                                uuid.uuid4().hex)
+        with open(tmp_new_filename, mode) as f:
+            writer(obj, f)
+        if fullpath.exists():
+            backup_path = fullpath.with_suffix(fullpath.suffix + ".backup")
+            if backup_path.exists():
+                backup_path.unlink()
+            fullpath.rename(backup_path)
+        tmp_new_filename.rename(fullpath)
+
+    def _read_object(self, reader, mode, path):
+        fullpath = self._workflow_root_dir / path
+        if fullpath.exists():
+            with open(fullpath, mode) as f:
+                return reader(f)
+        backup_path = fullpath.with_suffix(fullpath.suffix + ".backup")
+        if backup_path.exists():
+            with open(backup_path, mode) as f:
+                return reader(f)
+        raise FileNotFoundError(fullpath)
+
+    def file_exists(self, path: pathlib.Path) -> bool:
+        fullpath = self._workflow_root_dir / path
+        backup_path = fullpath.with_suffix(fullpath.suffix + ".backup")
+        return fullpath.exists() or backup_path.exists()
+
+    def directory_exists(self, path: pathlib.Path) -> bool:
+        fullpath = self._workflow_root_dir / path
+        return fullpath.exists()
+
+    def write_object_atomic(self, obj, path: pathlib.Path,
+                            overwrite=True) -> None:
+        self._write_atomic(ray.cloudpickle.dump, "wb", obj, path, overwrite)
+
+    def write_json_atomic(self, json_obj, path: pathlib.Path,
+                          overwrite=True) -> None:
+        self._write_atomic(json.dump, "w", json_obj, path, overwrite)
+
+    def read_object(self, path: pathlib.Path) -> Any:
+        return self._read_object(ray.cloudpickle.load, "rb", path)
+
+    def read_json(self, path: pathlib.Path) -> Any:
+        return self._read_object(json.load, "r", path)
+
+    def create_directory(self, path: pathlib.Path, parents=True,
+                         exist_ok=True) -> None:
+        dir_path = self._workflow_root_dir / path
+        dir_path.mkdir(parents=parents, exist_ok=exist_ok)
+
+
+_global_storage = Storage(
+    pathlib.Path(ray._private.utils.get_ray_temp_dir()) / "workflows")
```

Comment:
Please change this to use the present working directory as default (similar to metaflow).

Reply:
We should also print out the directory being used when starting a workflow.

Reply:
I was using the current working directory. But @iycheng favors a temp path more (like rllib/tune). @iycheng do you agree with the metaflow argument?

Reply:
Another reason to use the CWD is new users wouldn't know where the directory is by default. Putting it in the pwd makes it very obvious.

Reply:
My concern is that, if you put it in CWD, it means that when you resume the job you have to run it in that dir as well. 

And we have a job id, and if the user runs `resume(workflow_id)` in another dir, it'll fail.

Even the user doesn't know where the dir is, it still will work, right?

And rllib/rtune and even ray put spilled objects in other place by default.